### PR TITLE
Add eot token to ICL generate kwargs

### PR DESCRIPTION
**Pull request by bmosaicml**
# What does this PR do?

This PR sets the EOT early stoping criteria for ICL generation.

Tested F1 score generation without this change (run ` mpt-eval-DbThxP `). Sample output:

`' The book of Philippians was written to the church in the city of Philippi, which was located in present-day Greece.  The church in Philippi was founded by Paul the Apostle, and he wrote this letter while he was imprisoned in Rome.#include <stdio.h>\\n#include <stdlib.h>\\n#include <string.h>\\n#include'`

Benchmark scores: 
```
| Category   | Benchmark                        | Subtask   |     Score | Metric name                        | Number few shot   | Model                     |
|:-----------|:---------------------------------|:----------|----------:|:-----------------------------------|:------------------|:--------------------------|
|            | quac                             |           | 0.120712  | InContextLearningGenerationF1Score | 0-shot            | mosaicml/mpt-30b-instruct |
|            | natural_questions_closed         |           | 0.0832068 | InContextLearningGenerationF1Score | 0-shot            | mosaicml/mpt-30b-instruct |
|            | natural_questions_openbook_short |           | 0.23722   | InContextLearningGenerationF1Score | 0-shot            | mosaicml/mpt-30b-instruct |
|            | narrativeqa                      |           | 0.146517  | InContextLearningGenerationF1Score | 0-shot            | mosaicml/mpt-30b-instruct |
```

Output with correct eot token id: `' Abid Ali Neemuchwala'`

Run `mpt-eval-EYSsyE`

Benchmark scores:
```
| Category   | Benchmark                        | Subtask   |    Score | Metric name                        | Number few shot   | Model                     |
|:-----------|:---------------------------------|:----------|---------:|:-----------------------------------|:------------------|:--------------------------|
|            | quac                             |           | 0.217234 | InContextLearningGenerationF1Score | 0-shot            | mosaicml/mpt-30b-instruct |
|            | natural_questions_closed         |           | 0.189926 | InContextLearningGenerationF1Score | 0-shot            | mosaicml/mpt-30b-instruct |
|            | natural_questions_openbook_short |           | 0.706739 | InContextLearningGenerationF1Score | 0-shot            | mosaicml/mpt-30b-instruct |
|            | narrativeqa                      |           | 0.690028 | InContextLearningGenerationF1Score | 0-shot            | mosaicml/mpt-30b-instruct |
```

# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/datasets/in_context_learning_evaluation.py [hunk 1]
```diff
@@ -306,7 +306,8 @@ def collate_fn(self, data):
             'generation_length': self.max_answer_length,
             'generation_kwargs': {
                 'pad_token_id': self.pad_tok_id,
-                'use_cache': True
+                'use_cache': True,
+                'eos_token_id': self.tokenizer.eos_token_id
```
Comment (dakinggg) [inline composer/datasets/in_context_learning_evaluation.py:310]: assuming this is truly required, we should check somewhere that `eos_token_id` is not `None`, and error if it is

**Review by dakinggg — COMMENTED** [review]
Could you please update the PR description? I think it no longer reflects the state of the PR so I'm not sure if the PR implements what you wanted or not.

File: composer/datasets/in_context_learning_evaluation.py (hunk 1 above)
Comment (bmosaicml) [reply composer/datasets/in_context_learning_evaluation.py:310]: I don't think it shouldn't necessarily error because it's possible some models could have been trained to use a different token to indicate ends of sequences and we wouldn't want to completely rule them out, right?

Comment (dakinggg) [reply composer/datasets/in_context_learning_evaluation.py:310]: how would they specify that other than via eos token?

Comment (bmosaicml) [reply composer/datasets/in_context_learning_evaluation.py:310]: not sure, I was just assuming it's possible since couldn't you theoretically choose to put whatever token you want at the ends of your sequences? I could be wrong

File: composer/datasets/in_context_learning_evaluation.py [hunk 2]
```diff
@@ -154,6 +154,8 @@ def __init__(
         fewshot_random_seed: int,
         cot_delimiter: str = '',
     ):
+        if not hasattr(tokenizer, 'eos_token_id'):
```
Comment (dakinggg) [inline composer/datasets/in_context_learning_evaluation.py:157]: this isn't quite the right condition. it should be `if tokenizer.eos_token_id is None`. all tokenizers have the attr (https://github.com/huggingface/transformers/blob/224ab70969d1ac6c549f0beb3a8a71e2222e50f7/src/transformers/tokenization_utils_base.py#L795-L821). Could you add a simple test that if you try to construct the dataset with a tokenizer with a `None` eos token, the error is raised?

File: composer/datasets/in_context_learning_evaluation.py [hunk 3]
```diff
@@ -948,6 +951,8 @@ def __init__(
         top_p: Optional[float] = 0.95,
         top_k: Optional[int] = 40,
     ):
+        if not hasattr(tokenizer, 'eos_token_id'):
```
Comment (dakinggg) [inline composer/datasets/in_context_learning_evaluation.py:954]: same

File: tests/datasets/test_in_context_learning_datasets.py [hunk 4]
```diff
@@ -510,6 +510,35 @@ def test_qa_split_batch(tiny_opt_tokenizer, dataset_uri, tmp_path):
...
+def test_qa_task_dataloader_w_null_eos(dataset_uri, tiny_gpt2_tokenizer, tmp_path, num_fewshot, prompt_string):
+    pytest.importorskip('datasets')
+
+    local_data = os.path.join(os.path.dirname(__file__), 'local_data')
+
+    tokenizer = tiny_gpt2_tokenizer
+    dataset_uri = f'{local_data}/{dataset_uri}'
+    batch_size = 4
+    seqlen = 512
+    # empirical number from the small test dataset
```
Comment (dakinggg) [inline tests/datasets/test_in_context_learning_datasets.py:525]:
```suggestion
```

**Review by dakinggg — APPROVED** [review]
